crates/weaver_semconv/src/deprecated.rs

```diff
@@ -0,0 +1,341 @@
+                    Ok(Deprecated::Deprecated { note })
+                }
+                _ => Err(de::Error::missing_field("action")),
+            }
+        }
+    }
+
+    deserializer.deserialize_any(DeprecatedVisitor)
+}
+
+/// Custom deserialization function to handle both old and new formats for an optional field.
+pub fn deserialize_option_deprecated<'de, D>(
+    deserializer: D,
+) -> Result<Option<Deprecated>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    struct OptionDeprecatedVisitor;
+
+    impl<'de> Visitor<'de> for OptionDeprecatedVisitor {
+        type Value = Option<Deprecated>;
+
+        fn expecting(&self, formatter: &mut Formatter<'_>) -> fmt::Result {
+            formatter.write_str("a string, a map, or nothing for a deprecated field")
+        }
+
+        fn visit_unit<E>(self) -> Result<Self::Value, E>
+        where
+            E: de::Error,
+        {
+            // If we encounter an empty value (unit), we return None
+            Ok(None)
+        }
+
+        fn visit_none<E>(self) -> Result<Self::Value, E>
+        where
+            E: de::Error,
+        {
+            // Explicitly handle the None case (e.g., empty field)
+            Ok(None)
+        }
+
+        fn visit_some<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
+        where
+            D: Deserializer<'de>,
+        {
+            // Use the existing deserializer for Deprecated values and wrap the result in Some
+            let deprecated = deserialize_deprecated(deserializer)?;
+            Ok(Some(deprecated))
+        }
+    }
+
+    deserializer.deserialize_option(OptionDeprecatedVisitor)
+}
+
+/// Implements a human-readable display for the `Deprecated` enum.
+impl Display for Deprecated {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match self {
+            Deprecated::Renamed {
+                new_name: rename_to,
+                note,
+            } => {
+                if let Some(note) = note.as_ref() {
+                    write!(f, "{}", note)
+                } else {
+                    write!(f, "Replaced by `{}`.", rename_to)
+                }
+            }
+            Deprecated::Deprecated { note } => {
+                write!(f, "{}", note)
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[derive(Debug, Deserialize)]
+    struct Item {
+        #[serde(deserialize_with = "deserialize_option_deprecated", default)]
+        deprecated: Option<Deprecated>,
+    }
+
+    #[test]
+    fn test_deser_and_to_string() {
+        let yaml_data = r#"
+- deprecated: 'Replaced by `jvm.buffer.memory.used`.'
+- deprecated: 
+    action: deprecated
+    note: This attribute is no longer used.
+- deprecated: 
+    action: deprecated
+    note: Should no longer be used.
+- deprecated:
+    action: renamed
+    new_name: foo.unique_id
+    note: This field has been renamed for consistency.
```

Comment:
I thought we're going ahead with https://github.com/open-telemetry/semantic-conventions/issues/1426#issuecomment-2388775414

```yaml
- id: foo.id
  deprecated:
    action: renamed
    renamed_to: foo.unique_id
  note: we can write anything here
- id: foo.another_attribute
  deprecated: 
    action: removed
  note: another optional note
```

I.e. note is the attribute property.

This seems to be minimal and sufficient (we tend to remove notes from deprecated attributes, so it's totally fine to rewrite the original note and provide any deprecation info there).  I can see how `deprecated.note` could be nicer, but I'm suggesting to avoid adding it at least for now.

Reply:
I forgot this discussion... I'm fine with it (i.e. reusing the original note).

Reply:
I made this change. Now, the notes taken into account are those at the level of the object containing the deprecated field.

However, I renamed the `removed` action to `deprecated` because, in the diff, there is a distinction between something marked as `deprecated` and something that has disappeared from the registry. An attribute that has disappeared is marked as `removed` in the diff.

```
- id: foo.id
  deprecated:
    action: renamed
    renamed_to: foo.unique_id
  note: we can write anything here
- id: foo.another_attribute
  deprecated: 
    action: deprecated                 <-- was removed in your example
  note: another optional note
```